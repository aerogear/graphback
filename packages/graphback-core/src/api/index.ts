
export * from "./ContextCreator"
export * from "./ContextTypes"
export * from "./GraphbackCRUDGeneratorConfig"
export * from "./GraphbackOperationType"