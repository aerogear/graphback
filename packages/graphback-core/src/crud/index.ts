export * from "./mappingHelpers"
export * from "./GraphbackOperationType"
