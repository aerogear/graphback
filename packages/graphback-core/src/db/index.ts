export * from './defaultNameTransforms';
export * from './modelTableMapper';
export * from './dataMapper';
export * from './getPrimaryKey';
