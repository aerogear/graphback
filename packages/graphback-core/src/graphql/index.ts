
export * from "./GraphQLContextCreator"
export * from "./graphqlUtils"
export * from "./directives"