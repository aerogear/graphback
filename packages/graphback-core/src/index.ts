
//Top level API
export * from './crud'
export * from './plugin/GraphbackPlugin'
export * from './plugin/GraphbackPluginEngine'
export * from './plugin/GraphbackGlobalConfig'
export * from './plugin/GraphbackCRUDGeneratorConfig';
export * from './plugin/ModelDefinition'
export * from './plugin/getSelectedFieldsFromResolverInfo';
export * from './plugin/GraphbackCoreMetadata'

export * from './relationships/RelationshipMetadataBuilder';
export * from './relationships/relationshipHelpers';
export * from './annotations/DefaultValueAnnotation';

export * from './utils/printSchemaWithDirectives';
export * from './utils/metadataAnnotations';
export * from './utils/fieldTransformHelpers';
export * from './utils/isTransientField';

export * from './runtime';
export * from './db';

export * from './scalars'
