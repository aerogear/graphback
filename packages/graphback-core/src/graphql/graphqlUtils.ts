import { readFileSync } from 'fs';
import { sync } from 'glob';
import { join } from 'path';
import { GraphbackOperationType, InputModelTypeContext, INTERFACE_TYPE_DEFINITION, OBJECT_TYPE_DEFINITION, OBJECT_TYPE_EXTENSION } from '../api';

export const getFieldName = (typeName: string, action: GraphbackOperationType, plural: string = ''): string => {
  const upperCasedType = typeName.charAt(0).toUpperCase() + typeName.substr(1);

  return `${action}${upperCasedType}${plural}`
}

export const getTableName = (typeName: string): string => {
  return typeName.toLowerCase()
}

export const filterObjectTypes = (types: InputModelTypeContext[]) => types.filter((t: InputModelTypeContext) => t.kind === OBJECT_TYPE_DEFINITION);

export const filterInterfaceTypes = (types: InputModelTypeContext[]) => types.filter((t: InputModelTypeContext) => t.kind === INTERFACE_TYPE_DEFINITION);

export const filterObjectExtensions = (types: InputModelTypeContext[]) => types.filter((t: InputModelTypeContext) => t.kind === OBJECT_TYPE_EXTENSION);


