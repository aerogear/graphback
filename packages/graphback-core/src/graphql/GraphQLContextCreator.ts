import { parse, visit } from 'graphql';
import { InputModelTypeContext } from '../api/ContextTypes'
import { GraphbackCRUDGeneratorConfig } from "../api/GraphbackCRUDGeneratorConfig";
import { applyGeneratorDirectives } from './directives'
import { filterInterfaceTypes, filterObjectExtensions, filterObjectTypes } from './graphqlUtils';

import { InputContextCreator } from '../api';
import { inputTypeVisitor } from './InputTypeVisitor';



/**
 * Create input context to be reused for:
 * - schema
 * - various generators
 * - database creation
 */
export const graphQLInputContext: InputContextCreator = {

  createModelContext(schemaText: string, defaultConfig: GraphbackCRUDGeneratorConfig): InputModelTypeContext[] {
    const schema = applyGeneratorDirectives(schemaText)

    const astNode = parse(schema)
    const schemaDef = visit(astNode, { leave: inputTypeVisitor });
    const context = applyDefaultConfig(schemaDef, defaultConfig)
    const interfaces = filterInterfaceTypes(context)
    const finalTypes = applyExtensionsToOriginalTypes(context)

    return [...finalTypes, ...interfaces]

  }
}

// tslint:disable-next-line: no-any
function applyExtensionsToOriginalTypes(context: any) {
  const extendNodes = filterObjectExtensions(context)
  const filteredContext = filterObjectTypes(context)

  return filteredContext.map((t: InputModelTypeContext) => {
    const extendNode = extendNodes.find((node: InputModelTypeContext) => node.name === t.name);
    if (extendNode) {
      return {
        ...t,
        fields: [...t.fields, ...extendNode.fields],
        config: { ...t.config, ...extendNode.config },
        interfaces: [...t.interfaces, ...extendNode.interfaces]
      };
    }
    else {
      return t;
    }
  });
}

// tslint:disable-next-line: no-any
function applyDefaultConfig(schemaDef: any, defaultConfig: GraphbackCRUDGeneratorConfig) {
  return schemaDef.definitions.map((t: InputModelTypeContext) => {
    return {
      ...t,
      config: { ...defaultConfig, ...t.config }
    };
  });
}

