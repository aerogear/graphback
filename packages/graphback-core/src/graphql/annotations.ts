import { FieldDefinitionNode, TypeDefinitionNode } from 'graphql';
import { parseAnnotations } from 'graphql-annotations';
import { InputModelFieldAnnotations } from '../api';

export const parseFieldAnnotations = (node: FieldDefinitionNode): InputModelFieldAnnotations => {
    const fieldAnnotations: InputModelFieldAnnotations = {};

    const dbAnnotations: any = node.description ? parseAnnotations('db', String(node.description)) : undefined;

    if (!dbAnnotations) {
        return fieldAnnotations;
    }

    if (dbAnnotations.oneToOne) {
        fieldAnnotations.OneToOne = {
            field: dbAnnotations.oneToOne
        }
    }
    if (dbAnnotations.oneToMany) {
        fieldAnnotations.OneToMany = {
            field: dbAnnotations.oneToMany
        }
    }
    if (dbAnnotations.manyToOne) {
        fieldAnnotations.ManyToOne = {
            field: dbAnnotations.manyToOne
        }
    }
    if (dbAnnotations.manyToMany) {
        fieldAnnotations.ManyToMany = {
            field: dbAnnotations.manyToMany
        }
    }

    return fieldAnnotations;
}


export const parseTypeAnnotations = (node: TypeDefinitionNode) => {

    return node.description ? parseAnnotations('crud', String(node.description)) : undefined;
}
