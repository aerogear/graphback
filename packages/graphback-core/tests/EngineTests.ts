//tslint:disable-next-line: match-default-export-name no-implicit-dependencies
import { readFileSync } from 'fs';
import { buildSchema, print } from 'graphql';
import { GraphbackCoreMetadata, GraphbackPlugin, GraphbackPluginEngine } from '../src'

const schemaText = readFileSync(`${__dirname}/mock.graphql`, 'utf8')

/**
 * Testing engine
 */
class TestPlugin extends GraphbackPlugin {
  private generateCallback: any;

  //eslint-disable-next-line @typescript-eslint/explicit-member-accessibility
  constructor(generateCallback: any) {
    super();
    this.generateCallback = generateCallback;
  }

  public transformSchema(metadata: GraphbackCoreMetadata) {
    const schema = metadata.getSchema();
    schema.getQueryType().description = 'test';

    return schema;
  }

  /**
   * Create resources like files etc. for this plugin.
   * This method should write resouces to filesystem
   */
  public createResources(metadata: GraphbackCoreMetadata) {
    this.logError("I love")
    this.logWarning("High code coverage")
    this.generateCallback(metadata);
  }

  public getPluginName() {
    return "test"
  }
}

const context = {
  model: undefined
};

test('Test plugin engine', async () => {
  
  const crudMethods = {
    "create": true,
    "update": true,
    "findAll": true,
    "find": true,
    "delete": true,
  }

  let engine = new GraphbackPluginEngine(schemaText, { crudMethods: {} });
  engine = new GraphbackPluginEngine(buildSchema(schemaText), { crudMethods: crudMethods });

  expect.assertions(6);

  expect(engine.createResources).toThrow()
  const plugin = new TestPlugin((callbackModel: any) => {
    //eslint-disable-next-line dot-notation
    context['model'] = callbackModel;
    expect(callbackModel).toBeTruthy();
  })

  engine.registerPlugin(plugin, plugin, plugin)
  const model = engine.createResources();
  
  const printedModels = model.getModelDefinitions().map((element: any) => print(element.graphqlType.astNode))
  expect(printedModels).toMatchSnapshot();
  expect(model.getSchema().getQueryType().description === 'test').toBe(true);
});
