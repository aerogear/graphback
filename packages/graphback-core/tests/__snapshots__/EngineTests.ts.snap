// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test plugin engine 1`] = `
Array [
  "\\"\\"\\"
@model(delete: false)
@versioned
@delta
\\"\\"\\"
type Note {
  id: ID!
  title: String!
  description: String!
  \\"\\"\\"@oneToMany(field: 'noteComment')\\"\\"\\"
  comment: [Comment]!
}",
  "\\"\\"\\"@model(create: false)\\"\\"\\"
type Comment {
  id: ID!
  title: String!
  description: String!
  noteComment: Note!
}",
]
`;
